src/libsyntax/ext/expand.rs

```diff
@@ -218,21 +259,30 @@ fn expand_mac_invoc<T>(mac: ast::Mac, ident: Option<Ident>, attrs: Vec<ast::Attr
                     }
                 });
 
-                // DON'T mark before expansion.
-                fld.cx.insert_macro(ast::MacroDef {
+                let def = ast::MacroDef {
                     ident: ident,
                     id: ast::DUMMY_NODE_ID,
                     span: call_site,
                     imported_from: None,
                     use_locally: true,
                     body: marked_tts,
                     export: attr::contains_name(&attrs, "macro_export"),
-                    allow_internal_unstable: attr::contains_name(&attrs, "allow_internal_unstable"),
+                    allow_internal_unstable: attr::contains_name(&attrs,
+                                                                 "allow_internal_unstable"),
                     attrs: attrs,
-                });
+                };
 
-                // macro_rules! has a side effect but expands to nothing.
-                Some(Box::new(MacroScopePlaceholder))
+                // DON'T mark before expansion.
+                fld.cx.insert_macro(def.clone());
+
+                // macro_rules! has a side effect, but expands to nothing.
+                // If keep_macs is true, expands to a MacroDefExpn instead.
+                if fld.keep_macs {
+                    Some(Box::new(MacroDefExpn::new(def, path.clone())))
```

Comment:
With `MacEager::items`, this would be:
```rust
Some(MacEager::items(SmallVector::one(P(ast::Item {
    ident: def.ident,
    attrs: def.attrs.clone(),
    id: ast::DUMMY_NODE_ID,
    node: ast::ItemKind::Mac(ast::Mac {
        span: def.span,
        node: ast::Mac_ {
            path: path.clone(),
            tts: def.tts.clone(),
        }
    }),
    vis: ast::Visibility::Inherited,
    span: def.span,
}))))